### PR TITLE
Fixed #25809 -- Added BRINIndex support in django.contrib.postgres.

### PR DESCRIPTION
**Pull request by atombrella**
Preliminary work to add support for BRIN indexes in `django.contrib.postgres`.

I've added a link to Python sweetness in the documentation. 

The code is mostly adapted from `GinIndex`. I was tempted to include some more columns in the introspection to add some tests that `pages_per_range` is included properly.

### CODE REVIEW

File: docs/ref/contrib/postgres/indexes.txt [hunk 1]
```diff
@@ -24,3 +24,28 @@ available from the ``django.contrib.postgres.indexes`` module.
...
+    are only meant for system administration tasks, the functionality of the
+    module is not implemented.
+
+    To get the full advantage of using ``BRIN`` it is recommended to tweak the
+    index construction and consider the nature of the tables before using this
+    index. ``BRIN`` indexes take an optional input parameter ``pages_per_range``
+    for. The `Python sweetness`_ blog has a post with further details on the
+    ``pages_per_range`` parameter.
+
+    .. _Python sweetness: http://pythonsweetness.tumblr.com/post/119568339102/block-range-brin-indexes-in-postgresql-95
```
Comment (charettes) [inline docs/ref/contrib/postgres/indexes.txt]: We usually don't link to third party resource like this.

Comment (atombrella) [inline docs/ref/contrib/postgres/indexes.txt]: OK. My remark in the comment was intended as a question whether this was allowed.

File: django/contrib/postgres/indexes.py [hunk 2]
```diff
@@ -1,8 +1,26 @@
 from __future__ import unicode_literals
 
-from django.db.models import Index
+from django.db.models.indexes import Index
 
-__all__ = ['GinIndex']
+__all__ = ['BRINIndex', 'GinIndex']
+
+
+class BRINIndex(Index):
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: Seems like BrinIndex would be consistent with GinIndex.

File: django/contrib/postgres/indexes.py [hunk 3]
```diff
@@ -1,8 +1,26 @@
...
+    suffix = 'brin'
+
+    def __init__(self, *args, **kwargs):
+        self.pages_per_range = kwargs.pop('pages_per_range', None)
+        return super(BRINIndex, self).__init__(*args, **kwargs)
+
+    def create_sql(self, model, schema_editor):
+        sql = super(BRINIndex, self).create_sql(model, schema_editor, using=' USING brin')
+        if self.pages_per_range is not None:
+            sql += " WITH (pages_per_range = {})".format(self.pages_per_range)
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: You might append this to the `using` sql passed to `super()` to fix the test failure on Jenkins (due to usage of TABLESPACE).

File: django/contrib/postgres/indexes.py [hunk 4]
```diff
@@ -1,8 +1,26 @@
...
-from django.db.models import Index
+from django.db.models.indexes import Index
 
-__all__ = ['GinIndex']
+__all__ = ['BRINIndex', 'GinIndex']
+
+
+class BRINIndex(Index):
+    """
+    Support for the BRIN index introduced in PostgreSQL 9.5. There's an optional
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: I don't think it's important to mention PostgreSQL version details in the docstring.

File: docs/ref/contrib/postgres/indexes.txt [hunk 5]
```diff
@@ -24,3 +24,25 @@ available from the ``django.contrib.postgres.indexes`` module.
...
     operation.
+
+``BRINIndex``
+=============
+
+.. class:: BRINIndex(pages_per_range=None)
+
+    Creates a `BRIN index
+    <https://www.postgresql.org/docs/devel/static/brin-intro.html>`_. It is not
+    an optimal index to use unless your table is very large.  The `page inspect
```
Comment (timgraham) [inline docs/ref/contrib/postgres/indexes.txt]: I'm not sure if the Django docs is a good place to discuss the details of the index. Linking to the PostgreSQL docs seems better.

File: docs/releases/1.11.txt [hunk 6]
```diff
@@ -179,6 +179,10 @@ Minor features
 * The new :class:`~django.contrib.postgres.aggregates.JsonAgg` allows
   aggregating values as a JSON array.
 
+* Support for ``BRIN`` has been added in
```
Comment (timgraham) [inline docs/releases/1.11.txt]: combine this with the GinIndex note:
```
The new :class:`~django.contrib.postgres.indexes.GinIndex` and
:class:`~django.contrib.postgres.indexes.BrinIndex` classes allow
creating gin and brin indexes in the database.
```

File: tests/postgres_tests/test_indexes.py [hunk 7]
```diff
@@ -1,8 +1,33 @@
...
+from django.test import skipUnlessDBFeature
 
 from . import PostgreSQLTestCase
-from .models import IntegerArrayModel
+from .models import CharFieldModel, IntegerArrayModel
+
+
+class BRINIndexTests(PostgreSQLTestCase):
+
+    @skipUnlessDBFeature('has_brin_index_support')
```
Comment (timgraham) [inline tests/postgres_tests/test_indexes.py]: decorate the class instead

File: tests/postgres_tests/test_indexes.py [hunk 8]
```diff
@@ -1,8 +1,33 @@
...
+    def test_eq(self):
+        index = BRINIndex(fields=['title'])
+        same_index = BRINIndex(fields=['title'])
+        another_index = BRINIndex(fields=['author'])
+        self.assertEqual(index, same_index)
+        self.assertNotEqual(index, another_index)
+
+    @skipUnlessDBFeature('has_brin_index_support')
+    def test_deconstruction(self):
+        index = BRINIndex(fields=['title'], name='test_title_brin')
```
Comment (timgraham) [inline tests/postgres_tests/test_indexes.py]: Check with `pages_per_range`? (or maybe unneeded as above?)

File: tests/postgres_tests/test_indexes.py [hunk 9]
```diff
@@ -1,8 +1,33 @@
...
 
 from . import PostgreSQLTestCase
-from .models import IntegerArrayModel
+from .models import CharFieldModel, IntegerArrayModel
+
+
+class BRINIndexTests(PostgreSQLTestCase):
+
+    @skipUnlessDBFeature('has_brin_index_support')
+    def test_repr(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_indexes.py]: I'm not sure this is needed since class doesn't have a custom `__repr__()`, we're just testing the base class.

File: tests/postgres_tests/test_indexes.py [hunk 10]
```diff
@@ -1,8 +1,33 @@
...
+        index = BRINIndex(fields=['title'])
+        self.assertEqual(repr(index), "<BRINIndex: fields='title'>")
+
+    @skipUnlessDBFeature('has_brin_index_support')
+    def test_eq(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_indexes.py]: Check with `pages_per_range`? (or maybe unneeded as above?)

File: tests/postgres_tests/test_indexes.py (hunk 9 above)
Comment (atombrella) [inline tests/postgres_tests/test_indexes.py]: OK. This test should probably also be removed as well from the `GinIndex` tests.

File: tests/postgres_tests/test_indexes.py (hunk 10 above)
Comment (atombrella) [inline tests/postgres_tests/test_indexes.py]: Yes. I think this requires some addition to the introspection of the indexes? I've experimented with this. The key `expression` is `None` for the index.

File: tests/postgres_tests/test_indexes.py [hunk 11]
```diff
@@ -1,8 +1,32 @@
...
 
 from . import PostgreSQLTestCase
-from .models import IntegerArrayModel
+from .models import CharFieldModel, IntegerArrayModel
+
+
+@skipUnlessDBFeature('has_brin_index_support')
+class BRINIndexTests(PostgreSQLTestCase):
+
+    def test_eq(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_indexes.py]: If a custom `__eq__()` method isn't necesary, then it seems like these tests aren't needed.

File: django/contrib/postgres/indexes.py [hunk 12]
```diff
@@ -1,8 +1,33 @@
...
 
-__all__ = ['GinIndex']
+__all__ = ['BrinIndex', 'GinIndex']
+
+
+class BrinIndex(Index):
+    suffix = 'brin'
+
+    def __init__(self, *args, **kwargs):
+        self.pages_per_range = kwargs.pop('pages_per_range', 128)
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: Is there a problem putting this in the signature instead?

File: tests/postgres_tests/test_indexes.py [hunk 13]
```diff
@@ -1,8 +1,32 @@
...
+        another_index = BrinIndex(fields=['author'])
+        self.assertEqual(index, same_index)
+        self.assertNotEqual(index, another_index)
+
+    def test_not_eq(self):
+        index = BrinIndex(fields=['title'])
+        index_with_page_range = BrinIndex(fields=['title'], pages_per_range=16)
+        self.assertNotEqual(index, index_with_page_range)
+
+    def test_deconstruction(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_indexes.py]: A custom `deconstruct()` seems needed for `pages_per_range`.

File: django/contrib/postgres/indexes.py [hunk 14]
```diff
@@ -1,8 +1,33 @@
...
+        return super(BrinIndex, self).__init__(*args, **kwargs)
+
+    def __eq__(self, other):
+        return super(BrinIndex, self).__eq__(other) and self.pages_per_range == other.pages_per_range
+
+    def __repr__(self):
+        return "<%(name)s: %(field)s, pages_per_range=%(pages_per_range)s>" % {
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: Please test this method.

File: django/contrib/postgres/indexes.py [hunk 15]
```diff
@@ -1,8 +1,33 @@
...
+
+
+class BrinIndex(Index):
+    suffix = 'brin'
+
+    def __init__(self, *args, **kwargs):
+        self.pages_per_range = kwargs.pop('pages_per_range', 128)
+        return super(BrinIndex, self).__init__(*args, **kwargs)
+
+    def __eq__(self, other):
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: This shouldn't be needed once you add the custom `deconstruct()`.

File: docs/ref/contrib/postgres/indexes.txt [hunk 16]
```diff
@@ -24,3 +24,13 @@ available from the ``django.contrib.postgres.indexes`` module.
     PostgreSQL. You can install it using the
     :class:`~django.contrib.postgres.operations.BtreeGinExtension` migration
     operation.
+
+``BrinIndex``
```
Comment (timgraham) [inline docs/ref/contrib/postgres/indexes.txt]: Please alphabetize.

File: docs/ref/contrib/postgres/indexes.txt [hunk 17]
```diff
@@ -24,3 +24,13 @@ available from the ``django.contrib.postgres.indexes`` module.
...
+=============
+
+.. class:: BrinIndex(pages_per_range=None)
+
+    Creates a `Brin index
+    <https://www.postgresql.org/docs/devel/static/brin-intro.html>`_. For
+    performance considerations and use cases of the index, please consult the
+    documentation. Default value for ``pages_per_range`` is 128.
```
Comment (timgraham) [inline docs/ref/contrib/postgres/indexes.txt]: Would it be useful to explain how the default was chosen?

File: django/contrib/postgres/indexes.py [hunk 18]
```diff
@@ -1,8 +1,33 @@
...
+            'name': self.__class__.__name__,
+            'fields': self.fields,
+            'pages_per_range': self.pages_per_range,
+        }
+
+    def create_sql(self, model, schema_editor):
+        using = ' USING brin'
+        if self.pages_per_range is not None:
+            using += " WITH (pages_per_range = {})".format(self.pages_per_range)
+        sql = super(BrinIndex, self).create_sql(model, schema_editor, using=' USING brin')
```
Comment (holvianssi) [inline django/contrib/postgres/indexes.py]: Should be using=using?

File: django/contrib/postgres/indexes.py [hunk 19]
```diff
@@ -1,8 +1,49 @@
...
+
+    def __repr__(self):
+        return "<%(name)s: fields=%(fields)s, pages_per_range=%(pages_per_range)s>" % {
+            'name': self.__class__.__name__,
+            'fields': "'{}'".format(', '.join(self.fields)),
+            'pages_per_range': self.pages_per_range,
+        }
+
+    def deconstruct(self):
+        path = '%s.%s' % (self.__class__.__module__, self.__class__.__name__)
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: Can you use `super()` and add `pages_per_range` to the result?

File: django/contrib/postgres/indexes.py (hunk 18 above)
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: Maybe a `%(with)s` placeholder could be added to `sql_create_index`. Do you know or could you check if other databases have similar syntax?

Comment (atombrella) [inline django/contrib/postgres/indexes.py]: Yes. MySQL/MariaDB includes a lot of options after the ``USING``-clause. Oddly enough, there's no where-clause in the syntax. It's possible in all other three backends to limit the index to a subset of the table. I was tempted to suggest renaming extra to where, but because this is not implemented in MySQL/MariaDB, I won't.

 Oracle (for me, this is easier to read than the the official documentation: http://ss64.com/ora/index_c.html) also has some options before tablespace. I suppose Oracle users use all those options, and thus either may specify the raw SQL in ``create_sql``, or not include it in the django-code at all. 

SQLite is rather simple, and apart from the where-clause, the syntax is rather plain.

PostgreSQL of course has the with-clause, and at the end, the where-clause.

So, instead of naming the variable ``with``, the name ``specials`` (or ``options`') may be better? Anyway,  

This discussion brings to mind that the current implementation of ``GinIndex`` is incomplete, because the options ``fastupdate`` and ``gin_pending_list_limit`` are not available. I don't know if this falls into the category "not every database abstraction should be supported by Django" as the code is already in master.

**Comment by timgraham** [comment]
It looks to me like `WITH (pages_per_range = %d)` SQL clause is optional. Would it be better to default `pages_per_range=None` and omit the `WITH` clause to let PostgreSQL use its default in that case? I'm also thinking about if there was introspection support for that index, it seems like this behavior might be needed there.

File: django/db/models/indexes.py [hunk 20]
```diff
@@ -44,21 +44,27 @@ def check_name(self):
             self.name = 'D%s' % self.name[1:]
         return errors
 
-    def create_sql(self, model, schema_editor, using=''):
+    def index_parameters(self, model, schema_editor, using=''):
```
Comment (timgraham) [inline django/db/models/indexes.py]: Suggested name: `get_sql_create_params`. `BrinIndex` should override this function rather than overriding `create_sql`.

File: django/contrib/postgres/indexes.py [hunk 21]
```diff
@@ -1,8 +1,38 @@
...
+        path, args, kwargs = super(BrinIndex, self).deconstruct()
+        kwargs['pages_per_range'] = self.pages_per_range
+        return path, args, kwargs
+
+    def create_sql(self, model, schema_editor):
+        using = ' USING brin '
+        specials = 'WITH (pages_per_range = %d) ' % self.pages_per_range
+        parameters = self.index_parameters(model, schema_editor, using)
+        parameters['specials'] = specials
+        sql_create_index = "CREATE INDEX %(name)s ON %(table)s%(using)s (%(columns)s)%(specials)s%(extra)s"
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: This should probably be the default for postgresql's `schema_editor.sql_create_index`.

File: django/contrib/postgres/indexes.py [hunk 22]
```diff
@@ -1,8 +1,38 @@
...
+    def deconstruct(self):
+        path, args, kwargs = super(BrinIndex, self).deconstruct()
+        kwargs['pages_per_range'] = self.pages_per_range
+        return path, args, kwargs
+
+    def create_sql(self, model, schema_editor):
+        using = ' USING brin '
+        specials = 'WITH (pages_per_range = %d) ' % self.pages_per_range
+        parameters = self.index_parameters(model, schema_editor, using)
+        parameters['specials'] = specials
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: specials -> with?

File: docs/ref/contrib/postgres/indexes.txt [hunk 23]
```diff
@@ -9,6 +9,17 @@ PostgreSQL specific model indexes
 The following are PostgreSQL specific :doc:`indexes </ref/models/indexes>`
 available from the ``django.contrib.postgres.indexes`` module.
 
+``BrinIndex``
+=============
+
+.. class:: BrinIndex(pages_per_range=None)
+
+    Creates a `Brin index
```
Comment (timgraham) [inline docs/ref/contrib/postgres/indexes.txt]: Brin -> BRIN

File: docs/ref/contrib/postgres/indexes.txt [hunk 24]
```diff
@@ -9,6 +9,17 @@ PostgreSQL specific model indexes
...
+    <https://www.postgresql.org/docs/devel/static/brin-intro.html>`_. For
```
Comment (timgraham) [inline docs/ref/contrib/postgres/indexes.txt]: devel->current?

File: django/contrib/postgres/indexes.py [hunk 25]
```diff
@@ -1,8 +1,38 @@
...
 
-from django.db.models import Index
+from django.db.models.indexes import Index
 
-__all__ = ['GinIndex']
+__all__ = ['BrinIndex', 'GinIndex']
+
+
+class BrinIndex(Index):
+
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: chop blank line for consistency with `GinIndex`

File: django/contrib/postgres/indexes.py [hunk 26]
```diff
@@ -1,8 +1,38 @@
...
+    suffix = 'brin'
+
+    def __init__(self, fields=[], name=None, pages_per_range=128):
+        self.pages_per_range = pages_per_range
+        return super(BrinIndex, self).__init__(fields, name)
+
+    def __repr__(self):
+        return "<%(name)s: fields=%(fields)s, pages_per_range=%(pages_per_range)s>" % {
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: Use single quotes

File: docs/ref/contrib/postgres/indexes.txt (hunk 24 above)
Comment (atombrella) [inline docs/ref/contrib/postgres/indexes.txt]: Good. Will also make it consistent with the other links to PostgreSQL's documentation.

**Comment by atombrella** [comment]
I found the relevant column (``pg_class.reloptions``) for adding introspection, which returns strings like ``pages_per_range=128``. There's already some notes about ``get_constraints`` in the release notes. Where should I put documentation about this?

I'm adjusting the PR based on the comments.

File: django/contrib/postgres/indexes.py (hunk 21 above)
Comment (atombrella) [inline django/contrib/postgres/indexes.py]: There are quite a few places that would be affected by this. An alternative way to deal with this could be to introduce one more ``sql_create_index_sql_parameterized`` that has this ``with``-parameter. It would mean fewer ``try/except``-blocks etc. to handle the replacement.

**Comment by atombrella** [comment]
@timgraham Could you please trigger the build for Windows? 

I specifically did not squash the commits to make the changes a bit more visible. The discussion about the `sql_create_index` is tricky. I think it's better to use an additional template.

File: django/db/backends/postgresql/schema.py [hunk 27]
```diff
@@ -12,6 +12,7 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
     sql_set_sequence_max = "SELECT setval('%(sequence)s', MAX(%(column)s)) FROM %(table)s"
 
     sql_create_index = "CREATE INDEX %(name)s ON %(table)s%(using)s (%(columns)s)%(extra)s"
+    sql_create_index_parameterized = "CREATE INDEX %(name)s ON %(table)s%(using)s (%(columns)s)%(with)s%(extra)s"
```
Comment (MarkusH) [inline django/db/backends/postgresql/schema.py]: Can we put `'%(with)s'` into `sql_create_index` and make it default to an empty string? I'd prefer not having an extra SQL template for this if possible.

File: django/db/models/indexes.py [hunk 28]
```diff
@@ -44,21 +44,32 @@ def check_name(self):
...
+                'extra': tablespace_sql,
+                'with': '',
+                'using': using}
+
+    def create_sql(self, model, schema_editor, using='', parameters=None, sql_create_index=None):
+        sql_create_index = sql_create_index or schema_editor.sql_create_index
+        sql_parameters = parameters or self.get_sql_create_params(model, schema_editor, using)
+        try:
+            return sql_create_index % sql_parameters
+        except KeyError:
```
Comment (MarkusH) [inline django/db/models/indexes.py]: Can you elaborate on this except branch. Trying to figure out when this happens.

File: docs/releases/1.11.txt [hunk 29]
```diff
@@ -171,8 +171,9 @@ Minor features
   :class:`~django.contrib.postgres.aggregates.StringAgg` determines if
   concatenated values will be distinct.
 
-* The new :class:`~django.contrib.postgres.indexes.GinIndex` class allows
-  creating gin indexes in the database.
+* The new :class:`~django.contrib.postgres.indexes.GinIndex`  and
+  :class:`~django.contrib.postgres.indexes.BrinIndex` classes allow
+  creating gin and ``brin`` indexes in the database.
```
Comment (MarkusH) [inline docs/releases/1.11.txt]: make it `BRIN` and `GIN` please. That's the official spelling according to the PostgreSQL documentation.

File: django/db/models/indexes.py [hunk 30]
```diff
@@ -44,21 +44,32 @@ def check_name(self):
             self.name = 'D%s' % self.name[1:]
         return errors
 
-    def create_sql(self, model, schema_editor, using=''):
+    def get_sql_create_params(self, model, schema_editor, using=''):
```
Comment (MarkusH) [inline django/db/models/indexes.py]: `params` mostly refers to the parameters passed to `cursor.execute()` and are the values that are escaped and injected by the database (backend) at the very last before executing the query. Could you rename the method to something like `get_sql_create_template_values` if that makes sense?

File: docs/releases/1.11.txt (hunk 29 above)
Comment (atombrella) [inline docs/releases/1.11.txt]: The spelling of `gin` and `brin` was debated a bit in this PR already. It seems reasonably to use the same spelling as in the manual. Is changing gin -> GIN in this commit OK?

File: django/db/backends/postgresql/schema.py (hunk 27 above)
Comment (atombrella) [inline django/db/backends/postgresql/schema.py]: @timgraham also commented earlier on this. I'll try to remove the additional template. Initially, I introduced it to avoid dealing with the empty string for `with`.

File: django/db/models/indexes.py (hunk 28 above)
Comment (atombrella) [inline django/db/models/indexes.py]: Mainly because of the `with`-parameter. The chain to create the index somehow involves `_create_index_sql` in base/schema.py where `with` is not part of it. So in essence, the additional template string was an easy way to deal with it, but if possible I can look for ways to get rid of it.

Comment (MarkusH) [inline django/db/models/indexes.py]: Thanks. That's what I guessed. Adding a short comment on that would be fine, IMO.

File: docs/releases/1.11.txt (hunk 29 above)
Comment (MarkusH) [inline docs/releases/1.11.txt]: Yeah, feel free to update the spelling of both in here.

**Comment by auvipy** [comment]
@MarkusH can you approve it to be included in 1.11? or it's already missed the train?

**Comment by atombrella** [comment]
I've squashed  the commits, hoping this will ease the process before the feature freeze. Changes from both @MarkusH and @timgraham are both included; one of the the biggest headache seems to have been this `template`-string which ended up untouched.

File: docs/ref/schema-editor.txt [hunk 31]
```diff
@@ -184,3 +184,9 @@ accessed.
 
 This is useful when doing data migrations for :ref:`migrations with multiple
 databases <data-migrations-and-multiple-databases>`.
+
+Template strings
+----------------
+
+The ``SchemaEditor`` contains a number of template strings for creating and
```
Comment (MarkusH) [inline docs/ref/schema-editor.txt]: I'm trying to make sense of this sentence. It feels to me like a half-finished paragraph. :-/

File: django/db/models/indexes.py [hunk 32]
```diff
@@ -44,21 +44,24 @@ def check_name(self):
...
             for field, (field_name, order) in zip(fields, self.fields_orders)
         ]
-        return schema_editor.sql_create_index % {
-            'table': quote_name(model._meta.db_table),
-            'name': quote_name(self.name),
-            'columns': ', '.join(columns),
-            'using': using,
-            'extra': tablespace_sql,
-        }
+        return {'table': quote_name(model._meta.db_table),
```
Comment (MarkusH) [inline django/db/models/indexes.py]: Could you please follow the previous indentation style :)

File: django/contrib/postgres/indexes.py [hunk 33]
```diff
@@ -1,8 +1,37 @@
...
+
+    def deconstruct(self):
+        path, args, kwargs = super(BrinIndex, self).deconstruct()
+        kwargs['pages_per_range'] = self.pages_per_range
+        return path, args, kwargs
+
+    def get_sql_create_template_values(self, model, schema_editor, using):
+        parameters = super(BrinIndex, self).get_sql_create_template_values(model, schema_editor, using=' USING brin')
+        if self.pages_per_range is not None:
+            parameters['extra'] = ' WITH (pages_per_range={})'.format(self.pages_per_range) + parameters['extra']
```
Comment (MarkusH) [inline django/contrib/postgres/indexes.py]: Can you please run `pages_per_range` through `quote_value()` (internal function on the schema editor).

File: tests/postgres_tests/test_indexes.py [hunk 34]
```diff
@@ -1,8 +1,31 @@
...
+        index = BrinIndex(fields=['title'], name='test_title_brin')
+        path, args, kwargs = index.deconstruct()
+        self.assertEqual(path, 'django.contrib.postgres.indexes.BrinIndex')
+        self.assertEqual(args, ())
+        self.assertEqual(kwargs, {'fields': ['title'], 'name': 'test_title_brin', 'pages_per_range': None})
```
Comment (MarkusH) [inline tests/postgres_tests/test_indexes.py]: I'm missing a test where `pages_per_range` is defined. Can you also test that the value is escaped as per `quote_value()` (see comment above).

File: django/db/models/indexes.py (hunk 32 above)
Comment (atombrella) [inline django/db/models/indexes.py]: :)

File: docs/ref/schema-editor.txt (hunk 31 above)
Comment (atombrella) [inline docs/ref/schema-editor.txt]: There are quite a lot of those strings, `sql_create_index` being one. I could describe all of them, but I think that goes a bit beyond of the scope of this PR. I'll rework a bit; perhaps showing some examples is better.

Comment (MarkusH) [inline docs/ref/schema-editor.txt]: In other words, I don't think there's anything to document here.

Comment (timgraham) [inline docs/ref/schema-editor.txt]: Yea, changes in this document seems out of scope for this PR.

File: tests/postgres_tests/test_indexes.py (hunk 34 above)
Comment (atombrella) [inline tests/postgres_tests/test_indexes.py]: Added in a new test, along with a test to test for an invalid `pages_per_range` parameter. Probably why Github doesn't think any change took place.

File: django/contrib/postgres/indexes.py [hunk 35]
```diff
@@ -1,8 +1,40 @@
...
-__all__ = ['GinIndex']
+__all__ = ['BrinIndex', 'GinIndex']
+
+
+class BrinIndex(Index):
+    suffix = 'brin'
+
+    def __init__(self, fields=[], name=None, pages_per_range=None):
+        if pages_per_range is not None and pages_per_range <= 0:
+            raise TypeError('BRIN indexes require a positive integer for pages_per_range')
```
Comment (MarkusH) [inline django/contrib/postgres/indexes.py]: I'd go with `ValueError` and possibly add a check `isinstance(pages_per_range, int)`:

```python
if pages_per_range is not None and not (isinstance(pages_per_range, int) and pages_per_range > 0):
    raise ValueError('pages_per_range must be None or a positive integer for BRIN indexes')
```

File: tests/postgres_tests/test_indexes.py [hunk 36]
```diff
@@ -1,8 +1,43 @@
...
+        index = BrinIndex(fields=['title'], name='test_title_brin', pages_per_range=16)
+        path, args, kwargs = index.deconstruct()
+        self.assertEqual(path, 'django.contrib.postgres.indexes.BrinIndex')
+        self.assertEqual(args, ())
+        self.assertEqual(kwargs, {'fields': ['title'], 'name': 'test_title_brin', 'pages_per_range': 16})
+
+    def test_invalid_pages_per_range(self):
+        with self.assertRaises(TypeError):
+            BrinIndex(fields=['title'], name='test_title_brin', pages_per_range='Charles Babbage')
+            BrinIndex(fields=['title'], name='test_title_brin', pages_per_range=0)
```
Comment (MarkusH) [inline tests/postgres_tests/test_indexes.py]: You need to wrap the second instantiation in its own assertRaises to actually test it.

File: django/contrib/postgres/indexes.py [hunk 37]
```diff
@@ -1,8 +1,40 @@
 from __future__ import unicode_literals
 
-from django.db.models import Index
+from django.db.models.indexes import Index
```
Comment (akki) [inline django/contrib/postgres/indexes.py]: I am curious if there is any particular reason for this change.

Comment (atombrella) [inline django/contrib/postgres/indexes.py]: Not really. They're the same (`__init__.py` has two lines importing this). I can do a Refs-commit, if it's desired?

File: django/contrib/postgres/indexes.py [hunk 38]
```diff
@@ -1,8 +1,40 @@
...
-__all__ = ['GinIndex']
+__all__ = ['BrinIndex', 'GinIndex']
+
+
+class BrinIndex(Index):
+    suffix = 'brin'
+
+    def __init__(self, fields=[], name=None, pages_per_range=None):
+        if pages_per_range is not None and not (isinstance(pages_per_range, int) and pages_per_range > 0):
+            raise ValueError('pages_per_range must be None or a positive integer for BRIN indexes')
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: "for BRIN indexes" doesn't seem consistent with usual error messages.

File: django/contrib/postgres/indexes.py [hunk 39]
```diff
@@ -1,8 +1,40 @@
...
+        self.pages_per_range = pages_per_range
+        return super(BrinIndex, self).__init__(fields, name)
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: I don't think  `__init__()` returns?

File: tests/postgres_tests/test_indexes.py [hunk 40]
```diff
@@ -1,8 +1,45 @@
...
+
+    def test_deconstruction_with_pages_per_rank(self):
+        index = BrinIndex(fields=['title'], name='test_title_brin', pages_per_range=16)
+        path, args, kwargs = index.deconstruct()
+        self.assertEqual(path, 'django.contrib.postgres.indexes.BrinIndex')
+        self.assertEqual(args, ())
+        self.assertEqual(kwargs, {'fields': ['title'], 'name': 'test_title_brin', 'pages_per_range': 16})
+
+    def test_invalid_pages_per_range(self):
+        with self.assertRaises(ValueError):
```
Comment (timgraham) [inline tests/postgres_tests/test_indexes.py]: Use `assertRaisesMessage`.

File: tests/postgres_tests/test_indexes.py [hunk 41]
```diff
@@ -1,8 +1,45 @@
...
+        self.assertNotEqual(index, index_with_page_range)
+
+    def test_deconstruction(self):
+        index = BrinIndex(fields=['title'], name='test_title_brin')
+        path, args, kwargs = index.deconstruct()
+        self.assertEqual(path, 'django.contrib.postgres.indexes.BrinIndex')
+        self.assertEqual(args, ())
+        self.assertEqual(kwargs, {'fields': ['title'], 'name': 'test_title_brin', 'pages_per_range': None})
+
+    def test_deconstruction_with_pages_per_rank(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_indexes.py]: typo: `pages_per_range`

File: django/db/models/indexes.py [hunk 42]
```diff
@@ -44,22 +44,27 @@ def check_name(self):
...
-        return schema_editor.sql_create_index % {
+        return {
             'table': quote_name(model._meta.db_table),
             'name': quote_name(self.name),
             'columns': ', '.join(columns),
             'using': using,
             'extra': tablespace_sql,
         }
 
+    def create_sql(self, model, schema_editor, using='', parameters=None):
```
Comment (timgraham) [inline django/db/models/indexes.py]: I don't see the need for the `parameters=None` addition. Is it intentionally unused for now?

File: docs/ref/contrib/postgres/indexes.txt [hunk 43]
```diff
@@ -9,6 +9,16 @@ PostgreSQL specific model indexes
 The following are PostgreSQL specific :doc:`indexes </ref/models/indexes>`
 available from the ``django.contrib.postgres.indexes`` module.
 
+``BrinIndex``
+=============
+
+.. class:: BrinIndex(pages_per_range=None)
```
Comment (timgraham) [inline docs/ref/contrib/postgres/indexes.txt]: It seems `fields` and `name` should also be in the signature?

File: django/contrib/postgres/indexes.py [hunk 44]
```diff
@@ -1,8 +1,40 @@
...
 
-__all__ = ['GinIndex']
+__all__ = ['BrinIndex', 'GinIndex']
+
+
+class BrinIndex(Index):
+    suffix = 'brin'
+
+    def __init__(self, fields=[], name=None, pages_per_range=None):
+        if pages_per_range is not None and not (isinstance(pages_per_range, int) and pages_per_range > 0):
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: I'm not sure if `isinstance(pages_per_range, int)` is required. I think Python/Django doesn't do strict type checking like that in general.

Comment (atombrella) [inline django/contrib/postgres/indexes.py]: OK. https://github.com/django/django/compare/master...atombrella:refs_25809?expand=1 looks better. I guess a new PR is needed?
